.github/workflows/ci.yml

```diff
@@ -1,311 +1,177 @@
-            GIT_BRANCH=${{ env.GITHUB_REF_SHORT }}
-            GIT_COMMIT=${{ env.GITHUB_SHA_SHORT }}
-          # Only push if (a tag was passed in) or (we're building a non-master branch which isn't a PR)
-          push: ${{ github.event.inputs.tag != '' || (github.ref != 'refs/heads/master' && !contains(github.ref, 'refs/pull')) }}
-
-  # Create a new release if we're building a tag
-  create-release:
-    runs-on: ubuntu-latest
-    if: ${{ github.event.inputs.tag != '' }}
-    outputs:
-      upload_url: ${{ steps.create_release.outputs.upload_url }}
-    needs:
-      - dist
-      - build-publish
-      - build-publish-stretch
-
-    steps:
-      - name: Create Release
-        id: create_release
-        uses: actions/create-release@v1
-        env:
-          # Use custom secrets.GH_TOKEN instead of default secrets.GITHUB_TOKEN because the custom token will trigger the
-          # clarity-js-sdk-pr workflow. As events caused by default tokens do not trigger subsequent workflow runs to avoid loops.
-          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
-        with:
-          tag_name: ${{ github.event.inputs.tag || github.ref }}
-          release_name: Release ${{ github.event.inputs.tag || github.ref }}
-          draft: false
-          prerelease: true
+  # Create docker debian images
+  # Only run if:
+  #    - tag is provided
+  #    - AND
+  #        - this is the default branch
+  #        - AND Not a PR
+  docker-debian:
+    if: ${{ inputs.tag != '' && (github.ref == format('refs/heads/{0}', github.event.repository.default_branch) && !contains(github.ref, 'refs/pull')) }}
+    name: Docker Debian (Binary)
+    uses: stacks-network/stacks-blockchain/.github/workflows/image-build-debian-binary.yml@master
+    needs: github-release
+    with:
+      tag: ${{ inputs.tag }}
+      docker_platforms: linux/amd64, linux/amd64/v2, linux/amd64/v3
+      linux_version: debian
+      build_type: binary
+    secrets:
+      DOCKERHUB_USERNAME: ${{ secrets.DOCKERHUB_USERNAME }}
+      DOCKERHUB_PASSWORD: ${{ secrets.DOCKERHUB_PASSWORD }}
 
-  # Upload distributables to a new release if we're building a tag or a tag was passed in
-  upload-dist:
-    runs-on: ubuntu-latest
-    if: ${{ github.event.inputs.tag != '' }}
+  ###############################################
+  ## Build Branch/PR
+  ###############################################
+  # Create docker debian images
+  #   Only run if (a tag was NOT passed in) or (we're building a non-master branch which isn't a PR) or (this is a PR with no input tag)
+  # Only run if:
+  #    - tag is not provided
+  #    - OR
+  #        - non-default branch
+  #        - AND Not a PR
+  build-branch:
+    if: ${{ inputs.tag == '' || (github.ref != format('refs/heads/{0}', github.event.repository.default_branch) && !contains(github.ref, 'refs/pull')) }}
+    name: Docker Debian (Source)
+    uses: stacks-network/stacks-blockchain/.github/workflows/image-build-debian-source.yml@master
```

Comment:
Alternatively, you could make the `build-source` job run on all commits to build the binaries, [upload them to the workflow instance](https://github.com/actions/upload-artifact), then [download them](https://github.com/actions/download-artifact) to use them in the "binary" [docker images](https://github.com/stacks-network/stacks-blockchain/tree/806d99556432b9755d332d2526cf2e0d88b87847/.github/actions/dockerhub). Unless there's a reason you'd be unable to do this that I'm not seeing, you could eliminate duplicated code this way and ensure the same process is used between building PRs or a release.

[See the Clarinet CI workflow for an example if interested](https://github.com/hirosystems/clarinet/blob/develop/.github/workflows/ci.yaml)

Reply:
I really like that workflow....i'm going to have to take some time and see what i can lift from that. 
depending on timing though, it may have to be in the next iteration (alongside some composite actions perhaps?)

Reply:
resolving this for now -  I like the idea for the next iteration so i'm bookmarking that clarinet action